src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -486,22 +486,12 @@ MulticopterPositionControl::limit_altitude(vehicle_local_position_setpoint_s &se
 		return;
 	}
 
-	float altitude_above_home = -(_states.position(2) - _home_pos.z);
+	const float max_z = _home_pos.z + (-_vehicle_land_detected.alt_max);
 
-	if (altitude_above_home > _vehicle_land_detected.alt_max) {
-		// we are above maximum altitude
-		setpoint.z = -_vehicle_land_detected.alt_max +  _home_pos.z;
-		setpoint.vz = 0.0f;
-
-	} else if (setpoint.vz <= 0.0f) {
-		// we want to fly upwards: check if vehicle does not exceed altitude
-
-		float delta_p = _vehicle_land_detected.alt_max - altitude_above_home;
-
-		if (fabsf(setpoint.vz) * _dt > delta_p) {
-			setpoint.z = -_vehicle_land_detected.alt_max +  _home_pos.z;
-			setpoint.vz = 0.0f;
-		}
+	if (_states.position(2) < max_z) {
+		// we are above maximum altitude, don't allow further upwards flight
+		setpoint.z = max_z;
+		setpoint.vz = math::max(setpoint.vz, 0.f);
```

Comment:
So `vz` is positive to go down? Is that correct then?

Reply:
_Why not North West Up?_ 😜
Yes vz is positive to go down and `max(x,0)` makes sure it is positive and we can only go down.
But what I just saw is that a maximum altitude would translate to min_z because a highly negative altitude is a high altitude so I think I'll change the naming and comments again to make it more clear.

Reply:
Oh right!